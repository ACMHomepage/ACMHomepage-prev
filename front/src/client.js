import { ApolloClient, HttpLink, InMemoryCache } from '@apollo/client';

const link = new HttpLink({
  uri: 'http://localhost:3000/graphql',

  // Use explicit `window.fetch` so tha outgoing requests
  // are captured and deferred until the Service Worker is ready.
  fetch: (...args) => fetch(...args),
});

export const client = new ApolloClient({
  link,
  cache: new InMemoryCache(),
});
