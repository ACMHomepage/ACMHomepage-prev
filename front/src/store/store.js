import { configureStore } from '@reduxjs/toolkit';
import darkmodeReducer from './darkmodeSlice.js';

export default configureStore({
  reducer: {
    darkmode: darkmodeReducer,
  },
});
