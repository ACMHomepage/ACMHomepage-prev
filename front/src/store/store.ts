import { configureStore } from '@reduxjs/toolkit';
import darkmodeReducer from './darkmodeSlice';

export default configureStore({
  reducer: {
    darkmode: darkmodeReducer,
  },
});
