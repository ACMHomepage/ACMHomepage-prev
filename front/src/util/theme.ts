import type {
  Theme,
  ColorModesScale,
  ColorMode,
  ThemeUICSSObject,
  ThemeUIStyleObject,
} from 'theme-ui';
import fill from 'lodash/fill';
import get from 'lodash/get';
import isUndefined from 'lodash/isUndefined';

/******************************************************************************
 * Set the colors.
 *****************************************************************************/
type Color = ColorModesScale[string];
type ColorStyle = ThemeUICSSObject['color'];

interface Colors extends ColorModesScale {
  readonly green: Color;
  readonly gray: Color;
  modes: { dark: ColorMode };
}

const colors: Colors = {
  green: [
    '#F5FBF7', // 0
    '#E8F4EC', // 1
    '#D2E9DB', // 2
    '#B7D8C3', // 3
    '#9EC5AC', // 4
    '#78A288', // 5
    '#5A8068', // 6
    '#446953', // 7
    '#32533F', // 8
    '#244230', // 9
  ],
  gray: [
    '#F9FEFF', // 0
    '#DDE2E3', // 1
    '#C4CBCC', // 2
    '#A9B1B3', // 3
    '#909799', // 4
    '#747B7D', // 5
    '#5D6566', // 6
    '#43494B', // 7
    '#2D3233', // 8
    '#2E3233', // 9
  ],
  modes: {
    dark: {},
  },
};

colors.text = get(colors, 'green.9');
colors.modes.dark.text = get(colors, 'green.2');

colors.background = get(colors, 'green.0');
colors.modes.dark.background = get(colors, 'green.9');

colors.secondaryBackground = get(colors, 'green.2');
colors.modes.dark.secondaryBackground = get(colors, 'green.7');

// Fuck, now those are useless
colors.primary = 'white';
colors.modes.dark.primary = 'white';

colors.secondary = 'white';
colors.modes.dark.secondary = 'white';

interface SetColorConfig {
  setColorToBorderColor: boolean;
}

/**
 * setColor.
 *
 * @param color The text color and border color.
 * @param bg The background's color.
 * @param config Config:
 * - `setColorToBorderColor`: `true`(defalut) | `false`. Set param `color` to
 *   border's color or not.
 */
export const setColor = (
  color: ColorStyle,
  bg: ColorStyle,
  config?: SetColorConfig,
) => {
  if (isUndefined(config)) {
    config = {
      setColorToBorderColor: true,
    };
  }

  return {
    color,
    bg,
    borderColor: config.setColorToBorderColor ? color : undefined,
  };
};

/******************************************************************************
 * Set the break points.
 *****************************************************************************/
const __breakpointNameWithoutInitArray = ['sm', 'md', 'lg', 'xl'] as const;
if ('_' in __breakpointNameWithoutInitArray) {
  throw new Error(
    `Be sure that '_' cannot in the \`__breakpointNameWithoutInitArray\``,
  );
}

export type BreakpointNameWithoutInit =
  typeof __breakpointNameWithoutInitArray[number];
export type BreakpointName = BreakpointNameWithoutInit | '_';
export type BreakpointMapWithoutInit = {
  [size in BreakpointNameWithoutInit]: number;
};
export type BreakpointMap = {
  [size in BreakpointName]: number;
};

// Make breakpoints array and its name map(to breakpoints array's index).
// e.g. makeBreakPoints([['640px', 'sm'], ['780px', 'md']]) === [['640px',
// '780px'], {'sm': 0, 'md': 1}].
const makeBreakPoints = (
  breakpointsAndNames: Array<[string, BreakpointNameWithoutInit]>,
): [Array<string>, BreakpointMapWithoutInit] => {
  if (breakpointsAndNames.length !== __breakpointNameWithoutInitArray.length) {
    throw new Error(
      `Be sure that those names ${__breakpointNameWithoutInitArray},` +
        "are all in the argument `breakpointsAndNames`'s breakpoints' names",
    );
  }

  let result: [Array<string>, { [size: string]: number }] = [[], {}];
  let breakpointNames = new Set();
  for (let i = 0; i < breakpointsAndNames.length; i++) {
    let [size, name] = breakpointsAndNames[i];

    // make sure breakpoint name is unique.
    if (name in breakpointNames) {
      throw new Error("Be sure that breakpoint's name is unique.");
    }
    breakpointNames.add(name);

    result[0].push(size);
    result[1][name] = i;
  }
  return [result[0], result[1] as BreakpointMapWithoutInit];
};

const [breakpoints, breakpointNames] = makeBreakPoints([
  ['640px', 'sm'], // breakpoint 0
  ['780px', 'md'], // breakpoint 1
  ['1024px', 'lg'], // breakpoint 2
  ['1536px', 'xl'], // breakpoint 3
]);

// Make responsive value. (alias: mRV).
// e.g. If breakpointNames(should use `makeBreakPoints` to make) is {'sm': 0,
// 'md': 1, 'lg': 2, 'xl': 3}, then mRV<string>({_: '10px', 'md': '20px'}) ===
// ['10px', null, '20px', null, null].
export const makeResponsiveValue = <T>(
  argument: Partial<{ [size in BreakpointName]: T }>,
): (T | null)[] => {
  let result: (T | null)[] = fill(Array(breakpoints.length + 1), null);
  for (const s in argument) {
    const size = s as BreakpointName;
    const value = argument[size];
    if (isUndefined(value)) continue;
    if (size === '_') {
      result[0] = value;
      continue;
    }
    result[breakpointNames[size] + 1] = value;
  }
  return result;
};

/** Make responsive value. (alias: mRV). */
export const mRV = makeResponsiveValue;

/******************************************************************************
 * Set the radius size for border.
 *****************************************************************************/
const radii = {
  normal: '0.25rem',
  inf: '999999px',
};

interface SetBorderConfig {
  width?: ThemeUICSSObject['borderWidth'];
  radius?: ThemeUICSSObject['borderRadius'];
  style?: ThemeUICSSObject['borderStyle'];
}

/**
 * config:
 * - width: `<WIDTH>` | `'1px'`(defalut).
 * - radius: `<RADII>` | `'normal'`(defalut).
 * - style: `<BORDER-STYLE>` | `'solid'`(default).
 */
export const setBorder = (config?: SetBorderConfig) => {
  if (isUndefined(config)) config = {};
  if (isUndefined(config.width)) config.width = '1px';
  if (isUndefined(config.radius)) config.radius = 'normal';
  if (isUndefined(config.style)) config.style = 'solid';
  return {
    borderWidth: config.width,
    borderRadius: config.radius,
    borderStyle: config.style,
  };
};

/******************************************************************************
 * Set the weights, size for font.
 *****************************************************************************/
const fontWeights: Theme['fontWeights'] = {
  thin: 100,
  extralight: 200,
  light: 300,
  normal: 400,
  medium: 500,
  semibold: 600,
  bold: 700,
  extrabold: 800,
  black: 900,
};

fontWeights.h1 = fontWeights.extrabold;
fontWeights.h2 = fontWeights.bold;
fontWeights.h3 = fontWeights.semibold;
fontWeights.h4 = fontWeights.semibold;
fontWeights.h5 = fontWeights.medium;
fontWeights.h6 = fontWeights.medium;
fontWeights.body = fontWeights.normal;

const fontSizes: Theme['fontSizes'] = {
  xs: '0.75rem',
  sm: '0.875rem',
  base: '1rem',
  lg: '1.125rem',
  xl: '1.25rem',
  '2xl': '1.5rem',
  '3xl': '1.875rem',
  '4xl': '2.25rem',
  '5xl': '3rem',
  '6xl': '3.75rem',
  '7xl': '4.5rem',
  '8xl': '6rem',
  '9xl': '8rem',
};

type FontSizeName = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';

export const setFontSize = (name: FontSizeName): ThemeUICSSObject['fontSize'] => {
  switch (name) {
    case 'h1': return mRV({ _: '5xl', md: '6xl', lg: '7xl' });
    case 'h2': return mRV({ _: '4xl', md: '5xl', lg: '6xl' });
    case 'h3': return mRV({ _: '3xl', md: '4xl', lg: '5xl' });
    case 'h4': return mRV({ _: '2xl', md: '3xl', lg: '4xl' });
    case 'h5': return mRV({ _: 'xl', md: '2xl', lg: '3xl' });
    case 'h6': return mRV({ _: 'lg', md: 'xl', lg: '2xl' });
  }
}


/******************************************************************************
 * Set the flexbox.
 *****************************************************************************/
interface SetFlexConfig {
  center?: boolean;
  gap?: ThemeUICSSObject['gap'];
}

/**
 * setFlex.
 *
 * @param config Config the flexbox's style.
 * - `center`: `false`(default) | `true`. Set the content is put at the center
 *   or not.
 * - `gap`: `<LENGTH>` | `undefined`(default).
 */
export const setFlex = ({ center = false, gap }: SetFlexConfig) => {
  return {
    display: 'flex',
    alignItems: center ? 'center' : undefined,
    placeContent: center ? 'center' : undefined,
    gap,
  };
};

/******************************************************************************
 * Default theme.
 *****************************************************************************/
export default {
  config: {
    initialColorModeName: 'light',
  },
  colors,
  breakpoints,
  radii,
  fontWeights,
  fontSizes,
} as Theme;
