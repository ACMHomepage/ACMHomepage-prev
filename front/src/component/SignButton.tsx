import { LogIn, LogOut } from 'lucide-react';
import { useNavigate } from 'react-router-dom';

import { useSelector, useDispatch } from '../hooks';

import { selectState, AuthStateEnum, signOut } from '../store/authSlice';

import Button from './Button';

interface SignButtonProps {
  className?: string;
}

/**
 * SignButton. A button to go to log page.
 *
 * @param {SignButtonProps} props Holdes:
 * - className. `<string>`. To set its class.
 */

export default (props: SignButtonProps) => {
  const { className } = props;
  const navigate = useNavigate();
  const authState = useSelector(selectState);
  const dispatch = useDispatch();

  if (authState !== AuthStateEnum.LoggedWithInfo) {
    return (
      <Button
        sx={{ gap: '0.25rem' }}
        className={className}
        onClick={() => navigate('/sign')}
      >
        <LogIn size={16} />
        Sign in / Sign up
      </Button>
    );
  } else {
    // TODO: Let it can sign out.
    return (
      <Button
        sx={{ gap: '0.25rem' }}
        className={className}
        onClick={() => dispatch(signOut())}
      >
        <LogOut size={16} />
        Sign out
      </Button>
    );
  }
};
