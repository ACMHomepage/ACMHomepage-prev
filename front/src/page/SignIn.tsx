import { useEffect, useState } from 'react';
import { useNavigate, Link } from 'react-router-dom';
import { merge } from 'lodash';
import {
  border,
  bg,
  text,
  flexbox,
  font,
  spacing,
  size,
  layout,
} from '@acm-homepage/theme-shortcut';

import { useSelector } from '../hooks';
import { utilMainPart, boxSx } from '../config';
import { useSignIn, selectAuthState, AuthStateEnum } from '../store/authSlice';

import Header from '../component/Header';
import Button from '../component/Button';
import EmailInput from '../component/InputSet/EmailInput';
import PasswordInput from '../component/InputSet/PasswordInput';

export const URL = '/signin';

/**
 * Page `Sign`. Handle the sign in / sign on.
 */
export default () => {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const signIn = useSignIn();
  const authState = useSelector(selectAuthState);
  const navigator = useNavigate();

  useEffect(() => {
    if (authState === AuthStateEnum.LoggedWithInfo) {
      // Just go back.
      navigator(-1);
    }
  });

  return (
    <Header.Space sx={utilMainPart}>
      <div
        sx={merge(
          layout({ display: 'flex' }),
          flexbox({ gap: '1rem', dir: 'column' }),
          boxSx,
        )}
      >
        <div
          sx={merge(
            text({ align: 'center' }),
            font({ size: '4xl', weight: 'h3' }),
            spacing({ m: { b: '1.5rem' } }),
          )}
        >
          ACM Homepage
        </div>
        <EmailInput email={email} setEmail={setEmail} />
        <PasswordInput password={password} setPassword={setPassword} />
        <div sx={text({ align: 'right' })}>
          <a
            sx={merge(
              text({ col: 'link' }),
              font({ weight: 'link', size: 'sm' }),
            )}
          >
            Forget password?
          </a>
        </div>
        <Button
          sx={merge(
            border({ width: '2px', col: { _: 'bg-4', hv: 'bg-5' } }),
            size({ h: '2.5rem' }),
            font({ size: 'lg' }),
            spacing({ m: { t: '2rem' } }),
            text({ col: 'fg-0' }),
            bg({ col: { _: 'bg-0', hv: 'bg-2' } }),
          )}
          onClick={() => signIn(email, password)}
        >
          Sign in
        </Button>
      </div>
      <div sx={boxSx}>
        New user? Try to <Link to="/register">register</Link>.
      </div>
    </Header.Space>
  );
};
