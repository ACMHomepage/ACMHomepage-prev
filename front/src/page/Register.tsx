import { useEffect, useState } from 'react';
import { Lock, User, Mail, Eye } from 'lucide-react';
import { useNavigate, Link } from 'react-router-dom';

import { useSelector } from '../hooks';

import { utilMainPart, boxSx } from '../config';
import { setBorder, setColor, setFlex } from '../util/theme';
import { selectState, AuthStateEnum, useSignUp } from '../store/authSlice';

// Import components.
import Header from '../component/Header';
import Button from '../component/Button';
import Input from '../component/Input';

import { URL as signInUrl } from './SignIn';

export const URL = '/register';

/**
 * Page `Sign`. Handle the sign in / sign on.
 */
export default () => {
  const [name, setName] = useState('');
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const authState = useSelector(selectState);
  const navigator = useNavigate();

  const signUp = useSignUp();

  useEffect(() => {
    if (authState === AuthStateEnum.LoggedWithInfo) {
      // Just go back.
      navigator(-1);
    }
  });

  // TODO: Return error if the name, email or password is empty.
  // TODO: We do not handler if email is already in database.
  // TODO: We do not show the helpful message in different state.
  return (
    <Header.Space sx={{ ...utilMainPart }}>
      <div sx={{ ...boxSx, ...setFlex({ direction: 'column', gap: '1rem' }) }}>
        <div
          sx={{
            textAlign: 'center',
            fontSize: '4xl',
            fontWeight: 'h3',
            mb: '1.5rem',
          }}
        >
          ACM Homepage
        </div>
        <Input
          placeholder="User name"
          startIcon={User}
          value={[name, (event) => setName(event.target.value)]}
        />
        <Input
          placeholder="Email"
          startIcon={Mail}
          value={[email, (event) => setEmail(event.target.value)]}
        />
        <Input
          placeholder="Password"
          startIcon={Lock}
          endIcon={Eye}
          value={[password, (event) => setPassword(event.target.value)]}
        />
        <Button
          sx={{
            height: '2.5rem',
            fontSize: 'lg',
            mt: '2rem',
            ...setColor({ bg: 'bg-0', color: 'fg-0', hover: { bg: 'bg-2' } }),
            ...setBorder({ width: '2px', color: 'bg-4' }),
            '&:hover': {
              ...setBorder({ width: '2px', color: 'bg-5' }),
            },
          }}
          onClick={() => signUp(name, email, password)}
        >
          Resgister
        </Button>
      </div>
      <div sx={boxSx}>
        Already have an account? Try to <Link to={signInUrl}>sign in</Link>.
      </div>
    </Header.Space>
  );
};
