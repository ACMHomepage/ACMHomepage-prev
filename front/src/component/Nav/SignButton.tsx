import { LogIn, LogOut } from 'lucide-react';
import { useNavigate } from 'react-router-dom';

import { useSelector } from '../../hooks';
import { selectAuthState, AuthStateEnum } from '../../store/authSlice';
import { URL as signInUrl } from '../../page/SignIn';
import { useSignOut } from '../../api/auth';

import { useEffect } from 'react';

import styles from './styles/SignButton.module.scss';

/**
 * SignInOrRegisterButton. A button to go to sign in / register page.
 */
const SignInOrRegisterButton = (props: SignInOrRegisterProps) => {
  const navigate = useNavigate();

  return (
    <button className={styles.button} onClick={() => navigate(signInUrl)}>
      Sign in / Register
    </button>
  );
};

interface SignOutProps {
  className?: string;
}

/**
 * SignOutButton. A button to go to sign out.
 *
 * @param props - Holdes:
 * - className. `<string>`. To set its class.
 */
const SignOutButton = (props: SignOutProps) => {
  const { className } = props;

  const signOut = useSignOut();
  const authState = useSelector(selectAuthState);
  const navigate = useNavigate();

  useEffect(() => {
    if (authState === AuthStateEnum.Unlogged) {
      // If sign out, then go to the homepage.
      navigate('/');
    }
  }, [authState]);

  // TODO: Let it can sign out.
  return (
    <button className={className} onClick={signOut}>
      <LogOut size={16} />
      Sign out
    </button>
  );
};

interface SignButtonProps {
  className?: string;
}

/**
 * SignButton. A button to handle the event of signing in, registering and so
 * on.
 *
 * @param props - Holdes:
 * - className. `<string>`. To set its class.
 */
const SignButton = (props: SignButtonProps) => {
  const { className } = props;
  const authState = useSelector(selectAuthState);

  return authState !== AuthStateEnum.LoggedWithInfo ? (
    <SignInOrRegisterButton className={className} />
  ) : (
    <SignOutButton className={className} />
  );
};

export default SignButton;
