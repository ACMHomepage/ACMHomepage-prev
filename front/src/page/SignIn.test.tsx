// TODO: Move `*.test.ts` to folder `__tests__/`.
// TODO: Test that you cannot register two user with the same email.

import type { ReactNode } from 'react';
import { ThemeProvider } from 'theme-ui';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { ApolloProvider } from '@apollo/client';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import { Provider } from 'react-redux';
import { useLocation } from 'react-router-dom';

import theme from '../util/theme';
import store from '../store/store';
import { client } from '../client';

import SignIn, { URL as signInURL } from './SignIn';
import Register, { URL as registerURL } from './Register';
import SignButton from '../component/SignButton';

const Warpper = ({ children }: { children: ReactNode }) => (
  <Provider store={store}>
    <ApolloProvider client={client}>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </ApolloProvider>
  </Provider>
);

test('Make sure `Resgister` looks same as before.', () => {
  const { container } = render(
    <Warpper>
      <MemoryRouter>
        <SignIn />
      </MemoryRouter>
    </Warpper>,
  );

  expect(container).toMatchSnapshot();
});

test('The user can sign in', async () => {
  const Homepage = () => (
    <div>
      __Homepage__
      <SignButton />
    </div>
  );
  const _404 = () => {
    const location = useLocation();
    console.error(location);
    return <div>__404__</div>;
  };

  render(
    <Warpper>
      <MemoryRouter initialEntries={['/signin']}>
        <Routes>
          <Route path="/" element={<Homepage />} />
          <Route path={signInURL} element={<SignIn />} />
          <Route path={registerURL} element={<Register />} />
          <Route path="*" element={<_404 />} />
        </Routes>
      </MemoryRouter>
    </Warpper>,
  );

  expect(screen.queryByText('__Homepage__')).toBeNull();
  expect(screen.queryByText('__404__')).toBeNull();

  // Enter email and password, and submit.
  fireEvent.change(screen.getByPlaceholderText('Email'), {
    target: { value: 'not.peterlitszo@gmail.com' },
  });
  fireEvent.change(screen.getByPlaceholderText('Password'), {
    target: { value: 'admin123' },
  });
  fireEvent.click(screen.getByText('Sign in'));

  // Of course we cannot. We need register firstly.
  await waitFor(() => {
    expect(
      screen.getByText(
        'Error!! The email not.peterlitszo@gmail.com is not existed',
      ),
    ).toBeInTheDocument();
  });

  // Go to the register page.
  fireEvent.click(screen.getByText('register'));
  await waitFor(() => {
    expect(screen.getByText('Register')).toBeInTheDocument();
  });

  // Now we need enter infomation and submit to create new user.
  fireEvent.change(screen.getByPlaceholderText('Nickname'), {
    target: { value: 'PeterlitsZo' },
  });
  fireEvent.change(screen.getByPlaceholderText('Email'), {
    target: { value: 'not.peterlitszo@gmail.com' },
  });
  fireEvent.change(screen.getByPlaceholderText('Password'), {
    target: { value: 'admin123' },
  });
  fireEvent.click(screen.getByText('Register'));

  // We will at the homepage with a sign button.
  await waitFor(() => {
    expect(screen.getByText('__Homepage__')).toBeInTheDocument();
  });

  // But we need sign out.
  fireEvent.click(screen.getByText('Sign out'));

  // And now we need to sign in.
  await waitFor(() => {
    expect(screen.getByText('Sign in / Register')).toBeInTheDocument();
  });
  fireEvent.click(screen.getByText('Sign in / Register'));

  // Enter email and password, and submit.
  fireEvent.change(screen.getByPlaceholderText('Email'), {
    target: { value: 'not.peterlitszo@gmail.com' },
  });
  fireEvent.change(screen.getByPlaceholderText('Password'), {
    target: { value: 'admin123' },
  });
  fireEvent.click(screen.getByText('Sign in'));
});
