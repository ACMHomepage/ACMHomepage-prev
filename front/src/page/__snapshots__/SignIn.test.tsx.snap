// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Make sure \`Resgister\` looks same as before. 1`] = `
<div>
  <div
    class="css-1cd4a94-SignIn"
  >
    <div
      class="css-ovf630-SignIn"
    >
      <div
        class="css-4wu7wn-SignIn"
      >
        ACM Homepage
      </div>
      <label
        class="css-f5kwef"
      >
        <svg
          class="css-tzl219-EmailInput"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 4h16c1.1 0 2 .9 2 2v12c0 1.1-.9 2-2 2H4c-1.1 0-2-.9-2-2V6c0-1.1.9-2 2-2z"
          />
          <polyline
            points="22,6 12,13 2,6"
          />
        </svg>
        <input
          class="css-23obw9-EmailInput"
          placeholder="Email"
          value=""
        />
      </label>
      <label
        class="css-f5kwef"
      >
        <svg
          class="css-mahqe2-PasswordInput"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="11"
            rx="2"
            ry="2"
            width="18"
            x="3"
            y="11"
          />
          <path
            d="M7 11V7a5 5 0 0110 0v4"
          />
        </svg>
        <input
          class="css-ew39hz-PasswordInput"
          placeholder="Password"
          value=""
        />
        <svg
          class="css-1gd8vyt-PasswordInput"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1 12s4-8 11-8 11 8 11 8-4 8-11 8-11-8-11-8z"
          />
          <circle
            cx="12"
            cy="12"
            r="3"
          />
        </svg>
      </label>
      <div
        class="css-zgxne6-SignIn"
      >
        <a
          class="css-1bu445s-SignIn"
        >
          Forget password?
        </a>
      </div>
      <button
        class="css-3t1apv-SignIn"
      >
        Sign in
      </button>
    </div>
    <div
      class="css-1q2lkfp-SignIn"
    >
      New user? Try to 
      <a
        href="/register"
      >
        register
      </a>
      .
    </div>
  </div>
</div>
`;
